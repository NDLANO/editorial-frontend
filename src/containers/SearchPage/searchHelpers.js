import queryString from 'query-string';

export const converSearchStringToObject = (location, arrayFields = []) => {
  const searchLocation = queryString.parse(location ? location.search : '');

  return {
    ...searchLocation,
    ...arrayFields
      .map(field => ({
        [field]: searchLocation[field] ? searchLocation[field].split(',') : [],
      }))
      .reduce((result, item) => {
        const key = Object.keys(item)[0];
        return { ...result, [key]: item[key] };
      }, 0),
  };
};

export const convertSearchParam = value => {
  if (!value) {
    return undefined;
  } else if (Array.isArray(value)) {
    return value.length > 0 ? value.join(',') : undefined;
  } else if (Number.isInteger(value)) {
    return value;
  }
  return value.length > 0 ? value : undefined;
};
