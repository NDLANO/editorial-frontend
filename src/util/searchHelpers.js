/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

export const transformQuery = ({
  'resource-types': resourceTypes,
  ...rest
}) => {
  const query = { ...rest };

  if (resourceTypes === 'topic-article') {
    query['context-types'] = resourceTypes;
  } else if (resourceTypes) {
    query['resource-types'] = resourceTypes;
  }

  return query;
};
