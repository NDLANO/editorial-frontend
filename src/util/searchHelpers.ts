/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { UseQueryResult } from 'react-query';
import { SearchResultBase, SearchType } from '../interfaces';
import { useSearchAudio, useSearchSeries } from '../modules/audio/audioQueries';
import { useSearchConcepts } from '../modules/concept/conceptQueries';
import { useSearchImages } from '../modules/image/imageQueries';
import { useSearch } from '../modules/search/searchQueries';

export const transformQuery = ({ 'resource-types': resourceTypes, ...rest }: any) => {
  const query = { ...rest };

  if (resourceTypes === 'topic-article' || resourceTypes === 'frontpage-article') {
    query['article-types'] = resourceTypes;
  } else if (resourceTypes) {
    query['resource-types'] = resourceTypes;
  }
  if (query.users) {
    // in case of weird ID's starting with -
    query['users'] = `"${query['users']}"`;
  }

  return query;
};

type SearchHookType = (query: any) => UseQueryResult<SearchResultBase<any>>;

export const getSearchHookFromType = (type: SearchType): SearchHookType =>
  searchTypeToHookMapping[type] ?? useSearch;

const searchTypeToHookMapping: Record<SearchType, SearchHookType> = {
  audio: useSearchAudio,
  concept: useSearchConcepts,
  image: useSearchImages,
  'podcast-series': useSearchSeries,
  content: useSearch,
};
