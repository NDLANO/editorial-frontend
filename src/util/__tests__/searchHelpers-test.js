/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree. *
 */

import { transformQuery } from '../searchHelpers';

test('util/searchHelpers transformQuery', () => {
  const query = { page: '1', 'page-size': '10', sort: '-relevance' };
  expect(transformQuery(query)).toMatchSnapshot();
});

test('util/searchHelpers transformQuery', () => {
  const query = {
    page: '1',
    'page-size': '10',
    sort: '-relevance',
    'resource-types': 'urn:resourcetype:academicArticle',
  };
  expect(transformQuery(query)).toMatchSnapshot();
});

test('util/searchHelpers transformQuery', () => {
  const query = {
    page: '1',
    'page-size': '10',
    sort: '-relevance',
    'resource-types': 'topic-article',
  };
  expect(transformQuery(query)).toMatchSnapshot();
});
