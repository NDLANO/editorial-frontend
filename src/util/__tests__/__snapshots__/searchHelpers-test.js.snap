// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/searchHelpers transformQuery 1`] = `
{
  "page": "1",
  "page-size": "10",
  "sort": "-relevance",
}
`;

exports[`util/searchHelpers transformQuery 2`] = `
{
  "page": "1",
  "page-size": "10",
  "resource-types": "urn:resourcetype:academicArticle",
  "sort": "-relevance",
}
`;

exports[`util/searchHelpers transformQuery 3`] = `
{
  "article-types": "topic-article",
  "page": "1",
  "page-size": "10",
  "sort": "-relevance",
}
`;
